pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -1434,67 +1445,336 @@ private static void handleErrorInstance(String tableNameWithType, String segment
+  }
+
+  private static Map<String, Map<String, String>> getNextStrictReplicaGroupRoutingOnlyAssignment(
+      Map<String, Map<String, String>> currentAssignment, Map<String, Map<String, String>> targetAssignment,
+      int minAvailableReplicas, boolean lowDiskMode, int batchSizePerServer,
+      Object2IntOpenHashMap<String> segmentPartitionIdMap, PartitionIdFetcher partitionIdFetcher,
+      Logger tableRebalanceLogger) {
+    Map<String, Map<String, String>> nextAssignment = new TreeMap<>();
+    Map<String, Integer> numSegmentsToOffloadMap = getNumSegmentsToOffloadMap(currentAssignment, targetAssignment);
+    Map<Integer, Map<Set<String>, Map<String, Map<String, String>>>>
+        partitionIdToAssignedInstancesToCurrentAssignmentMap;
+    if (batchSizePerServer == RebalanceConfig.DISABLE_BATCH_SIZE_PER_SERVER) {
+      // Don't calculate the partition id to assigned instances to current assignment mapping if batching is disabled
+      // since we want to update the next assignment based on all partitions in this case. Use partitionId as 0
+      // and a dummy set for the assigned instances.
+      partitionIdToAssignedInstancesToCurrentAssignmentMap = new TreeMap<>();
+      partitionIdToAssignedInstancesToCurrentAssignmentMap.put(0, new HashMap<>());
+      partitionIdToAssignedInstancesToCurrentAssignmentMap.get(0).put(Set.of(""), currentAssignment);
+    } else {
+      partitionIdToAssignedInstancesToCurrentAssignmentMap =
+          getPartitionIdToAssignedInstancesToCurrentAssignmentMap(currentAssignment, segmentPartitionIdMap,
+              partitionIdFetcher);
+    }
+    Map<Pair<Set<String>, Set<String>>, Set<String>> assignmentMap = new HashMap<>();
+    Map<Set<String>, Set<String>> availableInstancesMap = new HashMap<>();
+
+    Map<String, Integer> serverToNumSegmentsAddedSoFar = new HashMap<>();
+    for (Map<Set<String>, Map<String, Map<String, String>>> assignedInstancesToCurrentAssignment
+        : partitionIdToAssignedInstancesToCurrentAssignmentMap.values()) {
+      boolean anyServerExhaustedBatchSize = false;
+      if (batchSizePerServer != RebalanceConfig.DISABLE_BATCH_SIZE_PER_SERVER) {
+        // Check if the servers of the first assignment for each unique set of assigned instances has any space left
+        // to move this partition. If so, let's mark the partitions as to be moved, otherwise we mark the partition
+        // as a whole as not moveable.
+        for (Map<String, Map<String, String>> curAssignment : assignedInstancesToCurrentAssignment.values()) {
+          Map.Entry<String, Map<String, String>> firstEntry = curAssignment.entrySet().iterator().next();
+          // All segments should be assigned to the same set of servers so it is enough to check for whether any server
+          // for one segment is above the limit or not
+          Map<String, String> firstEntryInstanceStateMap = firstEntry.getValue();
+          SingleSegmentAssignment firstAssignment =
```

Comment:
@Jackie-Jiang just FYI (same for where we calculate the strict replica group next assignment as well)

Here we just calculate the list of next servers and don't do the actual assignment. Still pass the original `numSegmentsToOffloadMap` and `assignmentMap`, but this does not update the `numSegmentsToOffloadMap` in this step, which means that the servers chosen may not correctly reflect based on offload map.

`assignmentMap` will be updated though and reused when we actually decide to do the assignment.

@Jackie-Jiang Do you see any concerns with this approach? I'm not sure how best to incorporate the segments to offload piece here without doing an actual assignment. Open to brainstorming ideas on how to handle this if you think it is a concern.